pkg/gatewayserver/io/ws/ws.go

```diff
@@ -304,7 +305,16 @@ func (s *srv) handleTraffic(w http.ResponseWriter, r *http.Request) (err error)
 					return err
 				}
 			case down := <-conn.Down():
-				concentratorTime, ok := conn.TimeFromTimestampTime(down.GetScheduled().Timestamp)
+				var (
+					concentratorTime scheduling.ConcentratorTime
+					ok               bool
+				)
+				// If absolute time is set, use that for computing xtime. If not, use the concentrator timestamp.
+				if down.GetScheduled().Time != nil {
+					concentratorTime, ok = conn.TimeFromGatewayTime(*ttnpb.StdTime(down.GetScheduled().Time))
```

Comment:
How sure are we that gateways have gateway time? From what I've seen they do, either via NTP or via time transfer.

The scheduler determines the concentrator time by gateway time otherwise by server time: https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.18/pkg/gatewayserver/scheduling/scheduler.go#L310-L321. That concentrator time is only used for duty-cycle enforcement and conflict detection.

What we really want to have here is the same concentrator time that the scheduler used too. With relative time scheduling, this is already in `Timestamp`. We explicitly don't set `Timestamp` with absolute time scheduling, see https://github.com/TheThingsNetwork/lorawan-stack/pull/1889/files#diff-9c87734043add0ade22bca71f2a6641c1d37830e1aae533fbc6066899c479d1b. Even if we had `Timestamp`, it's not the full concentrator time that we need here.

We cannot extend `TxSettings` with a `ConcentratorTime` field because downlink may not be scheduled by the Gateway Server (but by Packet Broker Agent).

So yeah, I don't know either what's the best solution here. I'm not against this approach, but there can be a discrepancy between the concentrator time here and the concentrator time that the scheduler determined. That may lead to a conflict or duty-cycle enforcement that is observed at the gateway. The only way to avoid that would be to extract that time determination logic from the scheduler to a method that can be called here.

Reply:
> We cannot extend `TxSettings` with a `ConcentratorTime` field because downlink may not be scheduled by the Gateway Server (but by Packet Broker Agent).

We could always fallback, right ? Use the concentrator time if available, otherwise check time and use the common function for deduction, otherwise fallback to timestamp.

Reply:
> Right and by `ConcentratorTime`, you mean the one from the emission right? I see how that could work actually.

Indeed. That's exactly the time the scheduler expects the emission to occur.

Reply:
Ah right. Yes indeed that would mean we can put the concentrator time in `TxSettings`.